cmd/snap-confine/mount-support-nvidia.c

```diff
@@ -294,44 +328,101 @@ static void sc_populate_libgl_with_hostfs_symlinks(const char *libgl_dir,
 	}
 }
 
+static void sc_overlay_init(const char *rootfs_dir)
+{
+	static char overlay_template[512];
+
+	sc_must_snprintf(overlay_template, sizeof(overlay_template), "%s/tmp/snap.nvidia_overlay_XXXXXX", rootfs_dir);
+	if (mkdtemp(overlay_template) == NULL) {
+		die("cannot create temporary directory %s for the nvidia overlay file system", overlay_template);
+	}
+
+	overlay_dir = overlay_template;
+
+	if (mount("none", overlay_dir, "tmpfs", MS_NODEV | MS_NOEXEC, NULL) != 0) {
+		die("cannot mount tmpfs at %s", overlay_dir);
+	}
+}
+
+static void sc_overlay_final(void)
+{
+	// Remount $tgt_dir (i.e. .../lib/gl) read only
+	debug("remounting overlay tmpfs as read-only %s", overlay_dir);
+	if (mount(NULL, overlay_dir, NULL, MS_REMOUNT | MS_BIND | MS_RDONLY, NULL) != 0) {
+		die("cannot remount %s as read-only", overlay_dir);
+	}
+}
+
+static int sc_overlay_mount(const char *target)
+{
+	char copy[512] = { 0 };
+
+	strncpy(copy, target, sizeof(copy) - 1);
+	const char *base_dir = basename(copy);
+	char overlay_upper[512] = { 0 };
+	char overlay_work[512] = { 0 };
+	char overlay_options[512] = { 0 };
+
+	sc_must_snprintf(overlay_upper, sizeof(overlay_upper), "%s/upper_%s", overlay_dir, base_dir);
+	sc_must_snprintf(overlay_work, sizeof(overlay_work), "%s/work_%s", overlay_dir, base_dir);
+	int res = mkdir(overlay_upper, 0755);
+	if (res != 0 && errno != EEXIST) {
+		die("cannot create overlay fs target %s", overlay_upper);
+	}
+	res = mkdir(overlay_work, 0755);
+	if (res != 0 && errno != EEXIST) {
+		die("cannot create overlay fs target %s", overlay_work);
+	}
+
+	res = mkdir(target, 0755);
+	if (res != 0 && errno != EEXIST) {
+		die("cannot create target %s", target);
+	}
+
+	sc_must_snprintf(overlay_options, sizeof(overlay_options), "lowerdir=%s,upperdir=%s,workdir=%s", target, overlay_upper, overlay_work);
+
+	if ((res = mount("overlay", target, "overlay", 0, overlay_options)) != 0) {
+		die("Unable to create overlay mount, target: %s, upper: %s, work: %s, options: %s, res: %d, errno: %d",
+				target, overlay_upper, overlay_work, overlay_options, res, errno);
+	}
+
+	return res;
+}
+
 static void sc_mkdir_and_mount_and_glob_files(const char *rootfs_dir,
-					      const char *source_dir[],
```

Comment:
Whilst I can understand the motivation for changing this (ie to keep the code simpler for callers), I think it is better to keep this as is to keep the interface flexible to allow for more than just 2 source directories to be specified.